File: sdks/insuretech-go-sdk/pkg/models/product.go
package models

import (
	"time"
)

// Product represents a product
type Product struct {
	BasePremiumCurrency string `json:"base_premium_currency,omitempty"`
	UpdatedAt time.Time `json:"updated_at,omitempty"`
	MaxSumInsured *Money `json:"max_sum_insured,omitempty"`
	CreatedAt time.Time `json:"created_at,omitempty"`
	DeletedAt time.Time `json:"deleted_at,omitempty"`
	AvailableRiders []*ProductsRider `json:"available_riders,omitempty"`
	MinSumInsuredCurrency string `json:"min_sum_insured_currency,omitempty"`
	MaxSumInsuredCurrency string `json:"max_sum_insured_currency,omitempty"`
	Category *ProductCategory `json:"category,omitempty"`
	MinTenureMonths int `json:"min_tenure_months,omitempty"`
	Exclusions []string `json:"exclusions,omitempty"`
	ProductId string `json:"product_id,omitempty"`
	Description string `json:"description,omitempty"`
	BasePremium *Money `json:"base_premium,omitempty"`
	MinSumInsured *Money `json:"min_sum_insured,omitempty"`
	MaxTenureMonths int `json:"max_tenure_months,omitempty"`
	Status *ProductsProductStatus `json:"status,omitempty"`
	CreatedBy string `json:"created_by,omitempty"`
	ProductAttributes string `json:"product_attributes,omitempty"`
	ProductCode string `json:"product_code,omitempty"`
	ProductName string `json:"product_name,omitempty"`
	PricingConfig *PricingConfig `json:"pricing_config,omitempty"`
	Plans []*ProductPlan `json:"plans,omitempty"`
}

File: backend/insurance_engine-old/src/InsuranceEngine.Products/Domain/Product.cs
using System;
using System.Collections.Generic;
using InsuranceEngine.Products.Domain.Enums;
using InsuranceEngine.SharedKernel.CQRS;
using InsuranceEngine.SharedKernel.Domain.ValueObjects;

namespace InsuranceEngine.Products.Domain;

public class Product
{
    public Guid Id { get; set; }
    public string ProductCode { get; set; } = string.Empty;
    public string ProductName { get; set; } = string.Empty;
    public string? Description { get; set; }
    public ProductCategory Category { get; set; }
    public ProductStatus Status { get; set; }

    // Money fields — stored as bigint (paisa)
    public long BasePremium { get; set; }
    public string BasePremiumCurrency { get; set; } = "BDT";
    public long MinSumInsured { get; set; }
    public string MinSumInsuredCurrency { get; set; } = "BDT";
    public long MaxSumInsured { get; set; }
    public string MaxSumInsuredCurrency { get; set; } = "BDT";

    public int MinTenureMonths { get; set; }
    public int MaxTenureMonths { get; set; }

    public List<string> Exclusions { get; set; } = new();
    public string? ProductAttributes { get; set; } // JSONB

    // Navigation properties
    public List<Rider> AvailableRiders { get; set; } = new();
    public List<ProductPlan> Plans { get; set; } = new();
    public List<RiskAssessmentQuestion> RiskAssessmentQuestions { get; set; } = new();
    public PricingConfig? PricingConfig { get; set; }

    // Audit
    public Guid CreatedBy { get; set; }
    public DateTime CreatedAt { get; set; }
    public DateTime UpdatedAt { get; set; }
    public DateTime? DeletedAt { get; set; }

    // --- Money convenience accessors ---
    public Money BasePremiumMoney
    {
        get => new(BasePremium, BasePremiumCurrency);
        set { BasePremium = value.Amount; BasePremiumCurrency = value.CurrencyCode; }
    }

    public Money MinSumInsuredMoney
    {
        get => new(MinSumInsured, MinSumInsuredCurrency);
        set { MinSumInsured = value.Amount; MinSumInsuredCurrency = value.CurrencyCode; }
    }

    public Money MaxSumInsuredMoney
    {
        get => new(MaxSumInsured, MaxSumInsuredCurrency);
        set { MaxSumInsured = value.Amount; MaxSumInsuredCurrency = value.CurrencyCode; }
    }

    // --- Status transition methods ---

    public Result Activate()
    {
        if (Status != ProductStatus.Draft && Status != ProductStatus.Inactive)
            return Result.Fail(Error.InvalidStateTransition(
                $"Cannot activate product in '{Status}' status. Only DRAFT or INACTIVE products can be activated."));

        Status = ProductStatus.Active;
        UpdatedAt = DateTime.UtcNow;
        return Result.Ok();
    }

    public Result Deactivate()
    {
        if (Status != ProductStatus.Active)
            return Result.Fail(Error.InvalidStateTransition(
                $"Cannot deactivate product in '{Status}' status. Only ACTIVE products can be deactivated."));

        Status = ProductStatus.Inactive;
        UpdatedAt = DateTime.UtcNow;
        return Result.Ok();
    }

    public Result Discontinue()
    {
        if (Status == ProductStatus.Discontinued)
            return Result.Fail(Error.InvalidStateTransition(
                "Product is already discontinued."));

        Status = ProductStatus.Discontinued;
        UpdatedAt = DateTime.UtcNow;
        return Result.Ok();
    }
}
